src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageClientMiddleLayer.cs

```diff
@@ -0,0 +1,215 @@
+            SVsServiceProvider serviceProvider)
+        {
+            if (joinableTaskContext is null)
+            {
+                throw new ArgumentNullException(nameof(joinableTaskContext));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (serviceProvider is null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            _joinableTaskFactory = joinableTaskContext.Factory;
+            _documentManager = documentManager;
+            _serviceProvider = serviceProvider;
+        }
+
+        public override bool CanHandle(string methodName)
+        {
+            return methodName == Methods.TextDocumentOnTypeFormattingName;
+        }
+
+        public override Task HandleNotificationAsync(string methodName, JToken methodParam, Func<JToken, Task> sendNotification)
+        {
+            return null;
+        }
+
+        public override async Task<JToken> HandleRequestAsync(string methodName, JToken methodParam, Func<JToken, Task<JToken>> sendRequest)
+        {
+            if (methodName == Methods.TextDocumentOnTypeFormattingName)
+            {
+                var emptyResult = JToken.FromObject(Array.Empty<TextEdit>());
+                var requestParams = methodParam.ToObject<DocumentOnTypeFormattingParams>();
+                if (requestParams.Options.OtherOptions == null)
+                {
+                    requestParams.Options.OtherOptions = new Dictionary<string, object>();
+                }
+
+                requestParams.Options.OtherOptions[LanguageServerConstants.ExpectsCursorPlaceholderKey] = true;
+                var token = JToken.FromObject(requestParams);
+                var result = await sendRequest(token);
+                var edits = result?.ToObject<TextEdit[]>();
+                if (edits == null)
+                {
+                    return emptyResult;
+                }
+
+                await _joinableTaskFactory.SwitchToMainThreadAsync();
+
+                if (!_documentManager.TryGetDocument(requestParams.TextDocument.Uri, out var documentSnapshot))
+                {
+                    return emptyResult;
+                }
+
+                Position cursorPosition = null;
+                var filteredEdits = new List<TextEdit>(edits.Length);
+                foreach (var edit in edits)
+                {
+                    if (edit.NewText.Contains(LanguageServerConstants.CursorPlaceholderString))
+                    {
+                        var newEdit = new TextEdit();
+                        newEdit.Range = edit.Range;
+                        newEdit.NewText = edit.NewText.Replace(LanguageServerConstants.CursorPlaceholderString, string.Empty);
+                        filteredEdits.Add(newEdit);
+
+                        cursorPosition = edit.Range.Start;
```

Comment:
This is assuming the placeholder is always at the beginning. Which it is currently; however, when WTE wants to support `attributeName="|"` it wont be because the edit will probably look like `="__placeholder__"`

Reply:
True. Will update